### ISSUE TITLE
MULTISET comparison doesn't work in Db2 with XML or JSONB emulation

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14469]
This test currently fails in Db2 with the XML emulation:

```sql
public void testMultisetInPredicates_EQ() throws Exception {
    testMultisetEmulations(ctx -> {
        assertEquals(1, ctx.fetchValue(
            selectCount()
           .from(TAuthor())
           .where(
               multiset(
                   select(TBook_ID()).from(TBook()).where(TBook_AUTHOR_ID().eq(TAuthor_ID()))
               )
               .eq(multiset(
                   select(val(2).as(TBook_ID()))
                   .union(
                   select(val(1).as(TBook_ID())))
               ))
           )
       ));
    });
}
```

The generated SQL is:

```sql
select count(*)
from "DB2INST1"."T_AUTHOR"
where (
  select xmlelement(
    name result,
    xmlagg(xmlelement(
      name record,
      xmlelement(name "v0", "DB2INST1"."T_BOOK"."ID")
    ))
  )
  from "DB2INST1"."T_BOOK"
  where "DB2INST1"."T_BOOK"."AUTHOR_ID" = "DB2INST1"."T_AUTHOR"."ID"
) = xmlserialize(content (
  select xmlelement(
    name result,
    xmlagg(xmlelement(
      name record,
      xmlelement(name "v0", "v0")
    ) order by "t"."v0")
  )
  from (
    select 2 "v0"
    from SYSIBM.DUAL
    union
    select 1 "ID"
    from SYSIBM.DUAL
  ) "t"
) as varchar(32672))
```

The exception is:

```
org.jooq.exception.DataAccessException: SQL [select count(*) from "DB2INST1"."T_AUTHOR" where (select xmlelement(name result, xmlagg(xmlelement(name record, xmlelement(name "v0", "DB2INST1"."T_BOOK"."ID")))) from "DB2INST1"."T_BOOK" where "DB2INST1"."T_BOOK"."AUTHOR_ID" = "DB2INST1"."T_AUTHOR"."ID") = xmlserialize(content (select xmlelement(name result, xmlagg(xmlelement(name record, xmlelement(name "v0", "v0")) order by "t"."v0")) from (select cast(? as integer) "v0" from SYSIBM.DUAL union select cast(? as integer) "ID" from SYSIBM.DUAL) "t") as varchar(32672)) -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT]; The data types of the operands for the operation "=" are not compatible or comparable.. SQLCODE=-401, SQLSTATE=42818, DRIVER=4.31.10
	at org.jooq_3.18.0-SNAPSHOT.DB2.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:3406)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:746)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:371)
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:301)
	at org.jooq.impl.AbstractResultQuery.fetchLazyNonAutoClosing(AbstractResultQuery.java:322)
	at org.jooq.impl.SelectImpl.fetchLazyNonAutoClosing(SelectImpl.java:2861)
	at org.jooq.impl.ResultQueryTrait.fetchOne(ResultQueryTrait.java:509)
	at org.jooq.impl.Tools.attach(Tools.java:1583)
	at org.jooq.impl.DefaultDSLContext.fetchOne(DefaultDSLContext.java:4706)
	at org.jooq.impl.DefaultDSLContext.lambda$55(DefaultDSLContext.java:4726)
	at org.jooq.impl.Tools.attach(Tools.java:1583)
	at org.jooq.impl.DefaultDSLContext.fetchValue(DefaultDSLContext.java:4726)
	at org.jooq.test.all.testcases.MultisetTests.lambda$373(MultisetTests.java:2487)
	at org.jooq.test.all.testcases.MultisetTests.lambda$467(MultisetTests.java:3065)
	at org.jooq.test.all.testcases.MultisetTests.lambda$470(MultisetTests.java:3083)
	at org.jooq.test.all.testcases.MultisetTests.testRecordLevelListenerResults(MultisetTests.java:3122)
	at org.jooq.test.all.testcases.MultisetTests.testRecordLevelListenerResults(MultisetTests.java:3110)
	at org.jooq.test.all.testcases.MultisetTests.testMultisetEmulations(MultisetTests.java:3083)
	at org.jooq.test.all.testcases.MultisetTests.testMultisetEmulations(MultisetTests.java:3065)
	at org.jooq.test.all.testcases.MultisetTests.testMultisetInPredicates_EQ(MultisetTests.java:2486)
	at org.jooq.test.jOOQAbstractTest.testMultisetInPredicates_EQ(jOOQAbstractTest.java:5780)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: com.ibm.db2.jcc.am.SqlSyntaxErrorException: The data types of the operands for the operation "=" are not compatible or comparable.. SQLCODE=-401, SQLSTATE=42818, DRIVER=4.31.10
	at com.ibm.db2.jcc.am.b7.a(b7.java:810)
	at com.ibm.db2.jcc.am.b7.a(b7.java:66)
	at com.ibm.db2.jcc.am.b7.a(b7.java:140)
	at com.ibm.db2.jcc.am.k_.c(k_.java:2844)
	at com.ibm.db2.jcc.am.k_.d(k_.java:2828)
	at com.ibm.db2.jcc.am.k_.a(k_.java:2254)
	at com.ibm.db2.jcc.am.la.a(la.java:8279)
	at com.ibm.db2.jcc.t4.ab.i(ab.java:204)
	at com.ibm.db2.jcc.t4.ab.b(ab.java:94)
	at com.ibm.db2.jcc.t4.p.a(p.java:32)
	at com.ibm.db2.jcc.t4.av.i(av.java:150)
	at com.ibm.db2.jcc.am.k_.al(k_.java:2223)
	at com.ibm.db2.jcc.am.la.bq(la.java:3763)
	at com.ibm.db2.jcc.am.la.a(la.java:4642)
	at com.ibm.db2.jcc.am.la.b(la.java:4215)
	at com.ibm.db2.jcc.am.la.bg(la.java:2886)
	at com.ibm.db2.jcc.am.la.execute(la.java:2861)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:219)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:219)
	at org.jooq.impl.Tools.executeStatementAndGetFirstResultSet(Tools.java:4691)
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:236)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:357)
	... 50 more
```

It seems only one side of the comparison is wrapped with `XMLSERIALIZE`. This would work as expected:

```sql
select count(*)
from "DB2INST1"."T_AUTHOR"
where xmlserialize(content (
  select xmlelement(
    name result,
    xmlagg(xmlelement(
      name record,
      xmlelement(name "v0", "DB2INST1"."T_BOOK"."ID")
    ))
  )
  from "DB2INST1"."T_BOOK"
  where "DB2INST1"."T_BOOK"."AUTHOR_ID" = "DB2INST1"."T_AUTHOR"."ID"
) as varchar(32672)) = xmlserialize(content (
  select xmlelement(
    name result,
    xmlagg(xmlelement(
      name record,
      xmlelement(name "v0", "v0")
    ) order by "t"."v0")
  )
  from (
    select 2 "v0"
    from SYSIBM.DUAL
    union
    select 1 "ID"
    from SYSIBM.DUAL
  ) "t"
) as varchar(32672))
```

**Comment by lukaseder** [comment]
This has already failed before jOOQ 3.18, but because of #12045, we hadn't integration tested this functionality, as correlated `MULTISET` subqueries weren't supported.

**Comment by lukaseder** [comment]
This is just an incorrect implementation of #12045, which explains why Oracle and PostgreSQL weren't affected.

**Comment by lukaseder** [comment]
The JSONB emulation also has a few flaws, which will be fixed (see also https://github.com/jOOQ/jOOQ/issues/14466).

**Comment by lukaseder** [comment]
As this all depends on #12045, which was fixed only for 3.18 and not backported, this fix here won't be backported either.
